docs/DafnyRef/UserGuide.md

```diff
@@ -156,14 +156,227 @@ exit code.
+  ensures 2 <= i
+{
+  var j := if !b then 3 else 1;
+  if b {
+    i := j;
+    assert 2 <= i; // This assertion might not hold
+    return;
+  }
+  i := 2;
+}
+```
+To debug why this assert might not hold, we need to _move this assert up_, which is similar to [_computing the weakest precondition_](https://en.wikipedia.org/wiki/Predicate_transformer_semantics#Weakest_preconditions).
+For example, if we have `x := Y; assert F(x);` and the `assert F(x);` might not hold, the weakest precondition for it to hold before `X := Y;` can be written as the assertion `assert F(Y);`, where we replace every occurence of `x` in `F(x)` into `Y`.
+Let's do it in our example:
+```dafny
+method FailingPostcondition(b: bool) returns (i: int)
+  ensures 2 <= i
+{
+  var j := if !b then 3 else 1;
+  if b {
+    assert 2 <= j; // This assertion might not hold
+    i := j;
+    assert 2 <= i;
+    return;
+  }
+  i := 2;
+}
+```
+Yay ! The assertion `assert 2 <= i;` is not proven wrong, which means that if we manage to prove `assert 2 <= j;`, it will work.
+Now, this assert should hold only if we are in this branch, so to _move the assert up_, we need to guard it.
+Just before the `if`, we can add the weakest precondition `assert b ==> (2 <= j)`:
+```dafny
+method FailingPostcondition(b: bool) returns (i: int)
+  ensures 2 <= i
+{
+  var j := if !b then 3 else 1;
+  assert b  ==>  2 <= j;  // This assertion might not hold
+  if b {
+    assert 2 <= j;
+    i := j;
+    assert 2 <= i;
+    return;
+  }
+  i := 2;
+}
+```
+Again, now the error is only on the topmost assert, which means that we are making progress.
+Now, either the error is obvious, or we can one more time replace `j` by its value and create the assert `assert b ==> ((if !b then 3 else 1) >= 2);`
+```dafny
+method FailingPostcondition(b: bool) returns (i: int)
+  ensures 2 <= i
+{
+  assert b  ==>  2 <= (if !b then 3 else 1);  // This assertion might not hold
+  var j := if !b then 3 else 1;
+  assert b  ==>  2 <= j;
+  if b {
+    assert 2 <= j;
+    i := j;
+    assert 2 <= i;
+    return;
+  }
+  i := 2;
+}
+```
+At this point, this is pure logic. We can simplify the assumption:
+```
+b ==>  2 <= (if !b then 3 else 1)
+!b ||  (if !b then 2 <= 3 else 2 <= 1)
+!b ||  (if !b then true else false)
+!b || !b;
+!b;
+```
+Now we can understand what went wrong: When b is true, all of these formulas above are false, this is why the Dafny verifier was not able to prove them.
+In the next section, we will explain how to `move asserts up` in certain useful patterns.
+
+#### 24.8.1.3. Failing asserts cases {#sec-failing-asserts-special-cases}
+
+This list is not exhaustive but can definitely be useful to provide the next step to figure out why Dafny could not prove an assertion.
+
+ Failing assert           | Suggested rewriting
+--------------------------|---------------------------------------
+ <br>`x := Y;`<br>`assert P[x];` | `assert P[Y];`<br>`x := Y;`<br>`assert P[x];`
```

Comment:
I would define the `P[x]` notation first, since you didn't use it earlier (and perhaps doing so in earlier sections where you say things like "replace every occurence of")